File: backend/app/schemas/files.py
from pydantic import BaseModel, Field
from typing import List

class FilePermissionRule(BaseModel):
    user_id: int
    can_view: bool = True
    can_edit: bool = True
    can_delete: bool = True

class FilePermissions(BaseModel):
    path: str
    owner_id: int
    rules: List[FilePermissionRule]

class FilePermissionsRequest(BaseModel):
    path: str
    owner_id: int = Field(..., gt=0)
    rules: List[FilePermissionRule]
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from app.schemas.sync import SyncDeviceInfo


class FileItem(BaseModel):
    name: str
    path: str
    size: int
    type: Literal["file", "directory"]
    modified_at: datetime
    owner_id: str | None = None
    mime_type: str | None = None
    file_id: int | None = None
    sync_info: list[SyncDeviceInfo] | None = None


class FileListResponse(BaseModel):
    files: list[FileItem]


class FileOperationResponse(BaseModel):
    message: str


class FileUploadResponse(FileOperationResponse):
    uploaded: int
    upload_ids: list[str] | None = None


class FolderCreateRequest(BaseModel):
    path: str | None = ""
    name: str


class RenameRequest(BaseModel):
    old_path: str
    new_name: str


class MoveRequest(BaseModel):
    source_path: str
    target_path: str


# ============================================================================
# Ownership Transfer Schemas
# ============================================================================

class OwnershipTransferRequest(BaseModel):
    """Request to transfer ownership of a file or directory."""
    path: str
    new_owner_id: int = Field(..., gt=0)
    recursive: bool = True
    conflict_strategy: Literal["rename", "skip", "overwrite"] = "rename"


class ConflictInfo(BaseModel):
    """Information about a naming conflict during transfer."""
    original_path: str
    resolved_path: str | None
    action: str  # "renamed", "skipped", "overwritten", "no_conflict"


class OwnershipTransferResponse(BaseModel):
    """Response from an ownership transfer operation."""
    success: bool
    message: str
    transferred_count: int = 0
    skipped_count: int = 0
    new_path: str | None = None
    conflicts: list[ConflictInfo] = []
    error: str | None = None


# ============================================================================
# Residency Enforcement Schemas
# ============================================================================

class ResidencyViolation(BaseModel):
    """Information about a file that violates residency rules."""
    path: str
    current_owner_id: int
    current_owner_username: str
    expected_directory: str
    actual_directory: str


class EnforceResidencyRequest(BaseModel):
    """Request to scan/fix residency violations."""
    dry_run: bool = True
    scope: str | None = None  # None = all users, or specific username


class EnforceResidencyResponse(BaseModel):
    """Response from a residency enforcement operation."""
    violations: list[ResidencyViolation] = []
    fixed_count: int = 0

File: client/src/api/files.ts
/**
 * API client for file operations (permissions, duplicates, ownership transfer, residency)
 */

import { apiClient, memoizedApiRequest } from '../lib/api';

// --- Duplicate Check API ---

export interface ExistingFileInfo {
  filename: string;
  size_bytes: number;
  modified_at: string;
  checksum: string | null;
}

export async function checkFilesExist(
  filenames: string[],
  targetPath: string,
): Promise<{ duplicates: ExistingFileInfo[] }> {
  const res = await apiClient.post('/api/files/check-exists', {
    filenames,
    target_path: targetPath,
  });
  return res.data;
}

// --- File Permissions API ---

export async function getFilePermissions(path: string) {
  // Memoized GET: Permissions werden für 60s gecached
  return memoizedApiRequest(`/api/files/permissions`, { path });
}

export async function setFilePermissions(data: {
  path: string;
  owner_id: number;
  rules: Array<{
    user_id: number;
    can_view: boolean;
    can_edit: boolean;
    can_delete: boolean;
  }>;
}) {
  const res = await apiClient.put(`/api/files/permissions`, data);
  return res.data;
}

// --- Ownership Transfer API ---

export interface ConflictInfo {
  original_path: string;
  resolved_path: string | null;
  action: string;
}

export interface OwnershipTransferRequest {
  path: string;
  new_owner_id: number;
  recursive?: boolean;
  conflict_strategy?: 'rename' | 'skip' | 'overwrite';
}

export interface OwnershipTransferResponse {
  success: boolean;
  message: string;
  transferred_count: number;
  skipped_count: number;
  new_path: string | null;
  conflicts: ConflictInfo[];
  error: string | null;
}

export async function transferOwnership(
  request: OwnershipTransferRequest
): Promise<OwnershipTransferResponse> {
  const res = await apiClient.post('/api/files/transfer-ownership', request);
  return res.data;
}

// --- Residency Enforcement API ---

export interface ResidencyViolation {
  path: string;
  current_owner_id: number;
  current_owner_username: string;
  expected_directory: string;
  actual_directory: string;
}

export interface EnforceResidencyRequest {
  dry_run?: boolean;
  scope?: string | null;
}

export interface EnforceResidencyResponse {
  violations: ResidencyViolation[];
  fixed_count: number;
}

export async function enforceResidency(
  request: EnforceResidencyRequest
): Promise<EnforceResidencyResponse> {
  const res = await apiClient.post('/api/files/enforce-residency', request);
  return res.data;
}
